datadog_sync/utils/base_resource.py

```diff
@@ -3,190 +3,207 @@
-                    print("{} resource ID {} diff: \n {}".format(self.resource_type, _id, pformat(diff)))
-            else:
-                print("Resource to be added {}: \n {}".format(self.resource_type, pformat(resource)))
+        try:
+            get_resp = self.get_resources(self.config.source_client)
+        except Exception as e:
+            self.config.logger.error(f"error while importing resources {self.resource_type}: {str(e)}")
+            return
 
-    def remove_non_nullable_attributes(self, resource):
-        if self.non_nullable_attr:
-            for key in self.non_nullable_attr:
-                k_list = key.split(".")
-                self.del_null_attr(k_list, resource)
+        futures = []
+        with thread_pool_executor(self.config.max_workers) as executor:
+            for r in get_resp:
+                if not self.filter(r):
+                    continue
+                futures.append(executor.submit(self.import_resource, r))
 
-    def apply_resources_sequentially(self, **kwargs):
-        resources = kwargs.get("resources") or self.source_resources
-        for _id, resource in resources.items():
+        for future in futures:
             try:
-                self.prepare_resource_and_apply(_id, resource, **kwargs)
-            except ResourceConnectionError:
-                # This should already be handled in connect_resource method
-                continue
+                future.result()
             except Exception as e:
-                self.logger.error(f"error while applying resource {self.resource_type}: {str(e)}")
+                self.config.logger.error(f"error while importing resource {self.resource_type}: {str(e)}")
+
+        write_resources_file(self.resource_type, SOURCE_ORIGIN, self.resource_config.source_resources)
+
+    def apply_resources(self) -> None:
+        max_workers = 1 if not self.resource_config.concurrent else self.config.max_workers
+
+        # Run pre-apply hook with the resources
+        try:
+            resources_list = self.pre_apply_hook(self.resource_config.source_resources)
+        except Exception as e:
+            self.config.logger.error(f"error while applying resources {self.resource_type}: {str(e)}")
+            return
+
+        futures = []
+        with thread_pool_executor(max_workers) as executor:
+            if resources_list:
+                for r_list in resources_list:
+                    for _id, resource in r_list.items():
+                        if not self.filter(resource):
+                            continue
+                        futures.append(executor.submit(self.apply_resource, _id, resource))
+                    wait(futures)
+            else:
+                for _id, resource in self.resource_config.source_resources.items():
+                    if not self.filter(resource):
+                        continue
+                    futures.append(executor.submit(self.apply_resource, _id, resource))
```

Comment:
```suggestion
            if not resources_list:
                resources_list = self.resource_config.source_resources
        
            for r_list in resources_list:
                for _id, resource in r_list.items():
                    if not self.filter(resource):
                        continue
                    futures.append(executor.submit(self.apply_resource, _id, resource))
```
Might be better this way, less duplicated code

Reply:
Same here, check the reply above. This is a workaround to ensure order is kept. E.g. simple and composite monitors